assets/src/components/cutoutOverlay.tsx

```diff
@@ -0,0 +1,86 @@
+import { LeafletMouseEvent } from "leaflet"
+import React, { useRef } from "react"
+import { useMapEvents } from "react-leaflet"
+
+/**
+ * Cutout Overlay Props
+ */
+type CutoutOverlayProps = {
+  /**
+   * Cutout Circle Reference for moving the cutout
+   */
+  circleTargetRef?: React.Ref<SVGCircleElement>
+  /**
+   * Containing box of the {@link CutoutOverlayProps.circleTargetRef}
+   *
+   * Used when calculating circle position from external coordinate systems
+   * such as mouse coordinates.
+   */
+  containerTargetRef?: React.Ref<SVGSVGElement>
+}
+
+/**
+ * Low Level cutout overlay, use map specific adapters
+ */
+export const CutoutOverlay = ({
+  circleTargetRef: circleTarget,
+  containerTargetRef: containerTarget,
+}: CutoutOverlayProps) => {
+  return (
+    <svg ref={containerTarget} className={"c-cutout-overlay"}>
+      <mask id="mask">
+        <rect className="c-cutout-overlay__mask-rect" />
+        <circle
+          className="c-cutout-overlay__mask-circle"
+          ref={circleTarget}
+        ></circle>
+      </mask>
+      <rect
+        // When the parent container (e.g., the map container) changes sizes,
+        // the parent SVG resizes, but that does not propagate down to this rect
+        // unless `width` and `height` are set inline.
+        width="100%"
+        height="100%"
+        className="c-cutout-overlay__rect"
+        mask="url(#mask)"
+      />
+    </svg>
+  )
+}
+
+/** {@see {@link CutoutOverlay.FollowMapMouseMove }} */
+const CutoutOverlayFollowMapMouseMove = () => {
+  const circleTarget = useRef<SVGCircleElement | null>(null)
+  const hoverTarget = useRef<SVGSVGElement | null>(null)
+
+  useMapEvents({
+    mousemove: (e: LeafletMouseEvent): void => {
+      if (circleTarget.current && hoverTarget.current) {
+        const { clientX: cursorScreenX, clientY: cursorScreenY } =
+          e.originalEvent
+        const { x: containerScreenX, y: containerScreenY } =
+          hoverTarget.current.getBoundingClientRect()
+
+        const containerOffsetX = cursorScreenX - containerScreenX
+        const containerOffsetY = cursorScreenY - containerScreenY
+
+        const maskPosition = `${containerOffsetX}px ${containerOffsetY}px`
+
+        circleTarget.current.style.translate = maskPosition
```

Comment:
**question:** I think handling this outside of the React component update flow (though if you do get a chance to testing out performance I'd be curious to see results). However, my one concern is if there is any way that React could rerender some components and clobber this `style.translate` change out from under us. That said, if that does happen then as soon as the user moves the pointer the mask will go back to the correct position, so maybe it's not such a big deal.

Reply:
I just tested this using state instead of refs, it doesn't completely get rid of ref's because you still need to account for the `containerRef` `<rect/>` for calculating offset within the container, and personally it feels slower/laggier. I am using a 165hz screen so it might be unnoticeable on 60hz, but I don't think it's worth it personally.

Additionally, it wasn't as easy to implement using state, and that could be contributing to the issues? Essentially, react does not seem to be updating `style.translate` when I pass it as a prop, and instead I had to use the `transform` prop and provide it the `transform` function in a string containing `transform(x px, y px)`. This is likely irrelevant, but I preferred using `style.translate`, as when I tried to access the `transform` property via javascript my IDE told me it was a readonly property (didn't dig further, might have been possible).

I'm going to leave it as is for now, I'll think on if I can distil any of what I found into useful inline comments.

Reply:
Sounds good! I don't know how many users we have with higher refresh rate monitors but if it's noticeably less responsive at 165 hz that's probably good evidence that we should keep the current approach rather than risk something that might become less responsive under certain conditions.